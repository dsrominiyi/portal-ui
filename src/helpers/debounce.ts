/**
 * Prevents an action from being called more that once every `wait` milliseconds
 * The initial triggering of the action will be delayed by `wait` milliseconds if `immediate` = false
 *
 * e.g. <button onClick={event => debounce(handleClick, 200)(event)} />
 */

let timeout: number | null;

const debounce = (action: (...args: any[]) => void, wait = 300, immediate = true) => (
  ...args: any[]
) => {
  const later = () => {
    timeout = null;
    if (!immediate) {
      action(...args);
    }
  };
  const callNow = immediate && !timeout;

  if (timeout) {
    clearTimeout(timeout);
  }

  timeout = (setTimeout as Window['setTimeout'])(later, wait);

  if (callNow) {
    action(...args);
  }
};

export default debounce;
