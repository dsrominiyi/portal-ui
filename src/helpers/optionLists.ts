export const maritalStatusOptions = [
  MaritalStatus.CommonLaw,
  MaritalStatus.Divorced,
  MaritalStatus.Married,
  MaritalStatus.Organisation,
  MaritalStatus.RegisteredCivilPartnership,
  MaritalStatus.Single,
  MaritalStatus.WidowWidower
];

export const consentTypeOptions = [
  MarketingConsentType.LegitimateInterest,
  MarketingConsentType.FaceToFace,
  MarketingConsentType.Phone,
  MarketingConsentType.Email,
  MarketingConsentType.Unsubscribe
];

export const contactMethodOptions = [
  ContactMethod.TextMessage,
  ContactMethod.Phone,
  ContactMethod.Email,
  ContactMethod.Post
];

export const additionalContactRelationshipOptions = [
  AdditionalContactRelationship.Spouse,
  AdditionalContactRelationship.Sibling,
  AdditionalContactRelationship.Parent,
  AdditionalContactRelationship.Child,
  AdditionalContactRelationship.Solicitor
];

export const contactTypeOptions = [
  ContactType.HomeNumber,
  ContactType.MobileNumber,
  ContactType.WorkNumber,
  ContactType.PersonalEmail,
  ContactType.WorkEmail
];

export const companyTypeStatusOptions = [
  CompanyTypeStatus.SoleProprietor,
  CompanyTypeStatus.Partnership,
  CompanyTypeStatus.Limited,
  CompanyTypeStatus.PublicLimited,
  CompanyTypeStatus.Other
];

export const otherCompanyTypeStatusOptions = [
  CompanyTypeStatus.AssociationClubSociety,
  CompanyTypeStatus.Charity,
  CompanyTypeStatus.Government,
  CompanyTypeStatus.LimitedLiabilityPartnership,
  CompanyTypeStatus.PartnershipLimited,
  CompanyTypeStatus.PartnershipOrdinary,
  CompanyTypeStatus.PrivateUnlimitedCompany,
  CompanyTypeStatus.TrustOrTrustees
];

export const referredByOptions = [
  'Practical Caravan',
  'Lead Linkedin',
  'C.V.R',
  'Autotrader Motorhome',
  'Yellow Pages',
  'Cross-selling',
  'Existing Business',
  'Lapsed Caravan',
  'Lapse CV',
  'Lapsed Home',
  'Ramsey Cricket Club',
  'Central Marketing',
  'Come Back Letter',
  'Fleet',
  'Stuarts Mailings',
  'Truck Tele',
  'Passing Trade',
  'Old Magazine Adverts',
  'Coachman Club',
  'Referral From NHF',
  'Recommended',
  'Vernon Property',
  'Insurecom Broker'
];

export const otherTitleOptions = [
  'Mx',
  'Master',
  'Brig.',
  'Bro.',
  'Capt.',
  'Chf.',
  'Col.',
  'Comdr.',
  'Comm.',
  'Cllr.',
  'Dame',
  'Dr.',
  'Exct.',
  'F/Lt.',
  'Gen.',
  'Hon.',
  'Judge',
  'Lady',
  'Lt.',
  'Lord',
  'Maj.',
  'Prof',
  'Rabbi',
  'Rev.',
  'Rt Rev',
  'Sir',
  'Sister',
  'Sq Ldr',
  'Rt Hon',
  'V. Rev',
  'W/Cdr'
];

export const individualAddressTypeOptions = ['Home', 'Correspondence'];

export const CompanyAddressTypeOptions = ['Correspondence', 'Trading', 'Registered'];

export const businessClassOptions = [
  'Accountancy',
  'Acoustic Engineer',
  'Actuarial Consultancy',
  'Acupuncture',
  'Addressing/Circularising Services',
  'Adjuster',
  'Advertising',
  'Aerial Erector',
  'Aerial Manufacturer',
  'Aerial Photography',
  'Aerial Supplier',
  'Aerial Survey',
  'Aerospace Industry',
  'Agricultural Engineer',
  'Agricultural Produce Distribution',
  'Agriculture',
  'Air Conditioning',
  'Air Transport'
];
