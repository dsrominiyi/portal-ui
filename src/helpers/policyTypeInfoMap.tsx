import React from 'react';

import PrivateCarIcon from '../assets/svg/icon-pt-private-car.svg';
import HouseholdIcon from '../assets/svg/icon-pt-household.svg';
import TravelIcon from '../assets/svg/icon-pt-travel.svg';
import CaravanIcon from '../assets/svg/icon-pt-caravan.svg';
import TruckIcon from '../assets/svg/icon-pt-truck.svg';
import CommercialCombinedIcon from '../assets/svg/icon-pt-commercial-combined.svg';
import PublicLiabilityIcon from '../assets/svg/icon-pt-public-liability.svg';
import CaterersIcon from '../assets/svg/icon-pt-caterers.svg';

export enum PolicyType {
  PrivateCar = 'pc',
  Household = 'hh',
  Travel = 'tv',
  Caravan = 'cn',
  Truck = 'cvt',
  CommercialCombined = 'cc',
  PublicLiability = 'pl',
  Caterers = 'ct'
}

export const policyTypeInfoMap = {
  [PolicyType.PrivateCar]: { icon: <PrivateCarIcon />, displayName: 'Private Car' },
  [PolicyType.Household]: { icon: <HouseholdIcon />, displayName: 'Household' },
  [PolicyType.Travel]: { icon: <TravelIcon />, displayName: 'Travel' },
  [PolicyType.Caravan]: { icon: <CaravanIcon />, displayName: 'Caravan' },
  [PolicyType.Truck]: { icon: <TruckIcon />, displayName: 'Truck' },
  [PolicyType.CommercialCombined]: {
    icon: <CommercialCombinedIcon />,
    displayName: 'Commercial Combined'
  },
  [PolicyType.PublicLiability]: {
    icon: <PublicLiabilityIcon />,
    displayName: 'Public Liability'
  },
  [PolicyType.Caterers]: { icon: <CaterersIcon />, displayName: 'Caterers' }
};
