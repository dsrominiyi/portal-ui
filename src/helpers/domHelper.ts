import debounce from './debounce';

export const preventBodyScroll = () => {
  const { scrollY } = window;
  document.body.style.position = 'fixed';
  document.body.style.top = document.body.style.top || `-${scrollY}px`;
};

export const allowBodyScroll = () => {
  const scrollY = document.body.style.top;
  document.body.style.position = '';
  document.body.style.top = '';
  return scrollY && window.scrollTo(0, -parseInt(scrollY, 10));
};

export const triggerOnEnter = (action: () => void, debounceWait?: number) => (
  event: React.KeyboardEvent<Element | HTMLElement>
) => {
  if (event.key === 'Enter') {
    debounce(action, debounceWait)();
  }
};
