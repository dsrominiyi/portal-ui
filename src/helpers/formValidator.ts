import { utc as moment } from 'moment';

import debounce from './debounce';
import { dateFieldValuesToDateString } from './utilFuncs';

/**
 * The formValidator helper functions supports a few different types of validation, these are
 * 'required', 'minLength', 'maxLength' and 'regex', where supplied values are simply used to
 * validate, additionally `predefinedRules` are made available, being made up of a range of rule
 * parts, these can then be imported for use in a form, and adjusted as individual needs require.
 */

export interface ValidationResult {
  isValid: boolean;
  message?: string;
}

export interface ValidationError {
  showError: boolean;
  message?: string;
}

export interface FormData {
  [formFieldKey: string]: FieldValue;
}

export enum RuleType {
  postcode = 'postcode'
}

export interface ValidationRule {
  fieldLabel: string;
  minLength?: number;
  maxLength?: number;
  required?: boolean;
  regex?: string | RegExp;
  isDate?: boolean;
  invalidValues?: string[];
  errorMessage?: string;
}

export interface ValidationRules {
  [formFieldKey: string]: ValidationRule;
}

export interface BatchValidationResult {
  [formFieldKey: string]: ValidationResult;
}

export interface FormErrors {
  [formFieldKey: string]: ValidationError;
}

export interface ValidationData {
  [formFieldKey: string]: ValidationResult;
}

export type ValidationStateArray = [
  BatchValidationResult,
  React.Dispatch<React.SetStateAction<BatchValidationResult>>
];

const isDateFieldValues = (value: FieldValue): value is DateFieldValues => {
  const objKeys = typeof value === 'object' && Object.keys(value as DateFieldValues);
  return objKeys && objKeys.includes('dd') && objKeys.includes('mm') && objKeys.includes('yyyy');
};

export const predefinedRules: ValidationRules = {
  postcode: {
    fieldLabel: 'postcode',
    minLength: 5,
    maxLength: 8,
    regex: /^[A-Z]{1,2}[0-9]{1,2}[A-Z]?(\s*[0-9][A-Z]{1,2})?$/,
    errorMessage: 'You must enter a valid UK postcode'
  },
  emailAddress: {
    fieldLabel: 'email address',
    maxLength: 254,
    regex: /^[^@]+@[^.]+\..+$/,
    errorMessage: 'You must enter a valid email address',
    required: true
  },
  phoneNumber: {
    fieldLabel: 'phone number',
    minLength: 7,
    maxLength: 50,
    regex: /^\+?[\d\s]+$/,
    errorMessage: 'You must enter a valid phone number',
    required: true
  },
  individualName: {
    fieldLabel: 'name',
    regex: /^[a-zA-Z-]+$/,
    minLength: 2,
    maxLength: 50
  },
  entityName: {
    fieldLabel: 'name',
    regex: /^[\w\s-]+$/,
    minLength: 2,
    maxLength: 50
  },
  registeredCoNumber: {
    fieldLabel: 'registration number',
    regex: /^([a-zA-Z]{2}[0-9]{6}|[0-9]{8})$/,
    errorMessage: 'You must enter a valid Companies House CRN',
    required: true
  }
};

export const validate = (rule: ValidationRule, value: FieldValue = ''): ValidationResult => {
  const {
    minLength,
    maxLength,
    required,
    regex,
    isDate,
    invalidValues,
    fieldLabel,
    errorMessage
  } = rule;

  const isBoolean = typeof value === 'boolean';
  const isArray = Array.isArray(value);
  const isEmptyArray = isArray && !(value as unknown[]).length;
  const nonRequiredAndEmpty = (!required && !value) || (!required && isEmptyArray);
  const requiredAndEmpty = (required && !value) || (required && isEmptyArray);

  if (isBoolean || nonRequiredAndEmpty) {
    return { isValid: true };
  }
  if (requiredAndEmpty) {
    return {
      isValid: false,
      message: errorMessage || `Please enter ${fieldLabel}`
    };
  }

  if (isDate) {
    const dateString = isDateFieldValues(value)
      ? dateFieldValuesToDateString(value)
      : (value as string);

    return (!dateString && !required) || moment(dateString, 'YYYY-MM-DD').isValid()
      ? { isValid: true }
      : {
          isValid: false,
          message: errorMessage || `${fieldLabel} should be a valid date`
        };
  }

  const stringValue = value as string;

  if (invalidValues && invalidValues.includes(stringValue)) {
    return {
      isValid: false,
      message:
        errorMessage || `${fieldLabel} cannot be one of the following: ${invalidValues.join(', ')}`
    };
  }
  if (minLength && minLength > stringValue.length) {
    return {
      isValid: false,
      message: errorMessage || `${fieldLabel} has a minimum length of ${minLength} characters`
    };
  }
  if (maxLength && maxLength < stringValue.length) {
    return {
      isValid: false,
      message: errorMessage || `${fieldLabel} has a maximum length of ${maxLength} characters`
    };
  }
  if (regex && !RegExp(regex).test(stringValue)) {
    return {
      isValid: false,
      message: errorMessage || `${fieldLabel} does not match the required format`
    };
  }

  return { isValid: true };
};

export const validateAll = (rules: ValidationRules, data: FormData): BatchValidationResult => {
  return Object.entries(data).reduce(
    (validationData, [formFieldKey, value]) => ({
      ...validationData,
      [formFieldKey]: validate(rules[formFieldKey], value)
    }),
    {}
  );
};

export const validateForm = (
  rules: ValidationRules,
  data: FormData,
  setValidationState: React.Dispatch<React.SetStateAction<BatchValidationResult>>
) => {
  const validationActions = () => {
    const validationResult = validateAll(rules, data);
    setValidationState(validationResult);
  };

  debounce(validationActions, 250, false)();
};

export const isFormValid = (validatedFields: BatchValidationResult) => {
  if (Object.keys(validatedFields).length === 0) {
    return false;
  }
  let allFieldsValid = true;

  for (const validatedField of Object.values(validatedFields)) {
    if (!validatedField.isValid) {
      allFieldsValid = false;
      break;
    }
  }

  return allFieldsValid;
};

export const buildErrors = (
  validationState: BatchValidationResult,
  submissionAttempted: boolean
): FormErrors => {
  return Object.entries(validationState).reduce(
    (errors, [formFieldKey, result]) => ({
      ...errors,
      [formFieldKey]: {
        showError: submissionAttempted && !result.isValid,
        message: result.message
      }
    }),
    {}
  );
};
