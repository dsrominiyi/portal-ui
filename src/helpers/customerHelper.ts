import { dateFieldValuesToDateString, dateStringToDateFieldValues } from './utilFuncs';
import { predefinedRules, ValidationRules } from './formValidator';

export const companyFormValidationRules: ValidationRules = {
  companyName: { ...predefinedRules.entityName, fieldLabel: 'company name', required: true },
  tradingName: { ...predefinedRules.entityName, fieldLabel: 'trading name' },
  typeStatus: { fieldLabel: 'company legal status', invalidValues: ['Other'], required: true },
  businessClass: { fieldLabel: 'business class', required: true },
  tradingSinceDate: { fieldLabel: 'trading since', isDate: true },
  parentCompany: { fieldLabel: 'is this company a subsidiary', required: true },
  parentCompanyName: { ...predefinedRules.entityName, fieldLabel: 'parent company name' },
  registeredCoNumber: predefinedRules.registeredCoNumber,
  contactDetails: { fieldLabel: 'contact details', required: true },
  addresses: { fieldLabel: 'addresses', required: true }
};

export const isIndividualCustomer = (customer: Customer): customer is Individual => {
  return typeof (customer as Individual).name !== 'undefined';
};

export const getIndividualFullName = ({ title, forename, middleName, surname }: IndividualName) => {
  return `${title} ${forename}${middleName && ` ${middleName}`} ${surname}`;
};

export const buildCompanyFromFormData = (data: NewCompanyData): Company => ({
  ...data,
  parentCompany: data.parentCompany || false,
  tradingSinceDate: dateFieldValuesToDateString(data.tradingSinceDate)
});

export const buildFormDataFromCompany = (data: Company): NewCompanyData => ({
  ...data,
  tradingSinceDate: data.tradingSinceDate
    ? dateStringToDateFieldValues(data.tradingSinceDate)
    : { dd: '', mm: '', yyyy: '' }
});
