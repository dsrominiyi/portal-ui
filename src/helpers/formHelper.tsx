import React, { ReactNode } from 'react';

import QuestionBox from '../components/QuestionBox';
import TextField, { Props as TextFieldProps } from '../components/TextField';
import AbiSearch, { Props as AbiSearchProps } from '../components/AbiSearch';
import Dropdown, { DropdownOption, Props as DropdownProps } from '../components/Dropdown';
import MultiChoice, { Props as MultiChoiceProps } from '../components/MultiChoice';
import DateFields, { Props as DateFieldsProps } from '../forms/common/DateFields';
import Lookup from '../components/Lookup';

import { ValidationError } from './formValidator';

import formStyles from '../forms/styles.scss';

interface InputConfig {
  type: InputType;
  label: string;
  buttonLabel?: string;
  secondaryButtonLabel?: string;
  value?: FieldValue;
  codeListId?: string;
  options?: DropdownOption[] | string[];
  attributes?: { [attr: string]: string };
  onChange?:
    | TextFieldProps['onChange']
    | AbiSearchProps['onChange']
    | DropdownProps['onChange']
    | DateFieldsProps['onChange']
    | MultiChoiceProps['onClick'];
  secondaryAction?: () => void;
}

export const buildInput = (config: InputConfig, fieldError?: ValidationError) => {
  let input: ReactNode;

  switch (config.type) {
    case InputType.Text:
      input = (
        <TextField
          key="0"
          additionalClasses={[formStyles.formTextField]}
          label={config.label}
          value={config.value as string}
          isInvalid={fieldError?.showError}
          onChange={config.onChange as TextFieldProps['onChange']}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.Lookup:
      input = (
        <Lookup
          key="0"
          label={config.label}
          buttonLabel={config.buttonLabel}
          secondaryButtonLabel={config.secondaryButtonLabel}
          value={config.value as string}
          isInvalid={fieldError?.showError}
          onChange={config.onChange as TextFieldProps['onChange']}
          secondaryAction={config.secondaryAction}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.AbiSearch:
      input = (
        <AbiSearch
          key="0"
          label={config.label}
          codeListId={config.codeListId || ''}
          onChange={config.onChange as AbiSearchProps['onChange']}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.Dropdown:
      input = (
        <Dropdown
          key="0"
          label={config.label}
          options={config.options!}
          selected={config.value as string}
          isInvalid={fieldError?.showError}
          onChange={config.onChange as DropdownProps['onChange']}
          attributes={config.attributes}
        />
      );
      break;
    case InputType.DateFields:
      input = (
        <DateFields
          key="0"
          values={config.value as DateFieldValues}
          onChange={config.onChange as DateFieldsProps['onChange']}
          testId={config.attributes && config.attributes['test-id']}
        />
      );
      break;
    case InputType.MultiChoice:
      input = (
        <MultiChoice
          key="0"
          options={config.options as string[]}
          selectedOption={config.value as string}
          onClick={config.onChange as MultiChoiceProps['onClick']}
          attributes={config.attributes}
        />
      );
      break;
  }

  return (
    <QuestionBox
      key={config.label}
      text={[InputType.DateFields, InputType.MultiChoice].includes(config.type) ? config.label : ''}
      hasError={fieldError?.showError}
      errorText={fieldError?.message}
      inputFields={[input]}
    />
  );
};
