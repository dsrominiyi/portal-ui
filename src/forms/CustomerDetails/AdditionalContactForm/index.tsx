import React, { FunctionComponent, useState, useEffect } from 'react';

import QuestionBox from '../../../components/QuestionBox';
import Dropdown from '../../../components/Dropdown';

import SelectTitle from '../SelectTitle';
import NameFields, { validationRules as nameValidationRules } from '../NameFields';
import ContactDetailFields, { getDetailValidationRule } from '../ContactDetailFields';
import ValidationErrorsNotification from '../../common/ValidationErrorsNotification';

import { additionalContactRelationshipOptions } from '../../../helpers/optionLists';
import { triggerOnEnter } from '../../../helpers/domHelper';
import {
  BatchValidationResult,
  ValidationStateArray,
  ValidationRules,
  isFormValid,
  validateForm,
  buildErrors
} from '../../../helpers/formValidator';

import formStyles from '../../styles.scss';

type UseStateArray = [AdditionalContact, React.Dispatch<React.SetStateAction<AdditionalContact>>];

const validationRules: ValidationRules = {
  ...nameValidationRules,
  title: {
    fieldLabel: 'title',
    required: true,
    invalidValues: ['Other']
  },
  relationship: {
    fieldLabel: 'relationship',
    required: true,
    errorMessage: 'Please confirm relationship'
  },
  type: { fieldLabel: 'contact detail type', required: true }
};

export const AdditionalContactForm: FunctionComponent<FormProps> = ({
  additionalClasses = [],
  initialValues,
  submissionAttempted: submissionAttemptedProp,
  onChange,
  onSubmit
}: FormProps) => {
  const [state, setState]: UseStateArray = useState(
    (initialValues as AdditionalContact) || {
      name: {
        title: '',
        forename: '',
        surname: ''
      },
      contactDetails: [
        {
          type: '' as ContactType,
          detail: ''
        }
      ],
      relationship: '' as AdditionalContactRelationship
    }
  );
  const [validationState, setValidationState]: ValidationStateArray = useState(
    {} as BatchValidationResult
  );
  const [submissionAttempted, setSubmissionAttempted] = useState(false);

  const { name, contactDetails, relationship } = state;
  const { title, forename, middleName, surname } = name;
  const { type, detail } = contactDetails[0];
  const canSubmit = isFormValid(validationState);

  useEffect(() => {
    const allValidationRules = { ...validationRules, detail: getDetailValidationRule(type) };
    const formData = {
      ...name,
      title,
      relationship,
      type,
      detail
    };
    validateForm(allValidationRules, formData, setValidationState);
    onChange(state, canSubmit);
  }, [state]);

  useEffect(() => {
    onChange(state, canSubmit);
  }, [validationState]);

  const submit = () => {
    setSubmissionAttempted(true);
    return canSubmit && onSubmit && onSubmit();
  };

  const hasAttemptedToSubmit = submissionAttemptedProp || submissionAttempted;
  const formErrors = buildErrors(validationState, hasAttemptedToSubmit);

  return (
    <div
      className={[formStyles.formSection, additionalClasses].join(' ')}
      onKeyUp={triggerOnEnter(submit)}
      test-id="additional-contact-form"
    >
      {hasAttemptedToSubmit && !canSubmit && <ValidationErrorsNotification />}
      <QuestionBox
        hasError={formErrors.relationship?.showError}
        errorText={formErrors.relationship?.message}
        inputFields={[
          <Dropdown
            key="0"
            label="Relationship"
            options={additionalContactRelationshipOptions}
            selected={relationship}
            isInvalid={formErrors.relationship?.showError}
            onChange={value => {
              setState({ ...state, relationship: value as AdditionalContactRelationship });
            }}
            attributes={{ 'test-id': 'select-relationship' }}
          />
        ]}
      />
      <SelectTitle
        title={title}
        hasError={formErrors.title?.showError}
        onChange={value => setState({ ...state, name: { ...name, title: value } })}
      />
      <NameFields
        forename={forename}
        middleName={middleName || ''}
        surname={surname}
        fieldErrors={{
          forename: formErrors.forename,
          middleName: formErrors.middleName,
          surname: formErrors.surname
        }}
        onChange={values => setState({ ...state, name: { ...values, title } })}
      />
      <ContactDetailFields
        type={type}
        detail={detail}
        fieldErrors={{
          type: formErrors.type,
          detail: formErrors.detail
        }}
        onChange={values => setState({ ...state, contactDetails: [values] })}
      />
    </div>
  );
};

AdditionalContactForm.displayName = 'AdditionalContactForm';
export default AdditionalContactForm;
