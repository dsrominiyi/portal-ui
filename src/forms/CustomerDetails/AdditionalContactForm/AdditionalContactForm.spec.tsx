import React from 'react';
import { shallow, mount, ShallowWrapper, ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

import AdditionalContactForm from '.';

import { isFormValid, buildErrors } from '../../../helpers/formValidator';

jest.mock('../../../helpers/formValidator');

let form: ShallowWrapper | ReactWrapper;
const onChange = jest.fn();
const onSubmit = jest.fn();
(isFormValid as jest.Mock).mockReturnValue(true);
(buildErrors as jest.Mock).mockReturnValue({ relationship: {}, title: {} });

const getRelationshipDropdown = () => {
  return (form
    .find('QuestionBox')
    .at(0)
    .props() as any).inputFields[0];
};

describe('AdditionalContactForm', () => {
  describe('snapshots', () => {
    beforeEach(() => {
      form = shallow(<AdditionalContactForm onChange={onChange} onSubmit={onSubmit} />);
    });

    test('AdditionalContactForm matches snapshot', () => {
      expect(form).toMatchSnapshot();
    });

    test('AdditionalContactForm with initial values matches snapshot', () => {
      (form as ShallowWrapper).setProps({
        initialValues: {
          name: { title: 'Mr', forename: 'Robert', middleName: 'Dave', surname: 'Smith' },
          contactDetails: [{ type: ContactType.HomeNumber, detail: '08589392900' }],
          relationship: AdditionalContactRelationship.Sibling
        }
      });

      expect(form).toMatchSnapshot();
    });

    test('AdditionalContactForm without form errors matches snapshot', () => {
      (buildErrors as jest.Mock).mockReturnValueOnce({});
      form = shallow(<AdditionalContactForm onChange={onChange} onSubmit={onSubmit} />);

      expect(form).toMatchSnapshot();
    });
  });

  describe('functionality', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      form = mount(<AdditionalContactForm onChange={onChange} onSubmit={onSubmit} />);
    });

    test('onSubmit is called when the Enter is pressed', () => {
      form.simulate('keyup', { key: 'Enter' });

      expect(onSubmit).toHaveBeenCalled();
    });

    test('onSubmit is not called when the Enter is pressed if the form is invalid', () => {
      (isFormValid as jest.Mock).mockReturnValueOnce(false);
      form.simulate('keyup', { key: 'Enter' });

      expect(onSubmit).not.toHaveBeenCalled();
    });

    test('ValidationErrorsNotification is rendered when the Enter is pressed but the form is invalid', () => {
      (isFormValid as jest.Mock).mockReturnValueOnce(false);
      (form as ReactWrapper).setProps({ submissionAttempted: true });
      form.simulate('keyup', { key: 'Enter' });

      expect(form.find('ValidationErrorsNotification').length).toBe(1);
    });

    test('onChange is called with the form state when any field is updated', () => {
      const relationship = AdditionalContactRelationship.Spouse;
      const relationshipDropdown = getRelationshipDropdown();

      act(() => {
        relationshipDropdown.props.onChange(relationship);
      });

      expect(onChange).toHaveBeenCalledWith(
        {
          name: {
            title: '',
            forename: '',
            surname: ''
          },
          contactDetails: [
            {
              type: '' as ContactType,
              detail: ''
            }
          ],
          relationship
        },
        true
      );

      form.update();
      const title = 'Mr';

      act(() => {
        (form.find('SelectTitle').props() as any).onChange(title);
      });

      expect(onChange).toHaveBeenCalledWith(
        {
          name: {
            title,
            forename: '',
            surname: ''
          },
          contactDetails: [
            {
              type: '' as ContactType,
              detail: ''
            }
          ],
          relationship
        },
        true
      );

      form.update();
      const forename = 'Jim';
      const middleName = 'Jack';
      const surname = 'Jones';
      const nameFields = { forename, middleName, surname };

      act(() => {
        (form.find('NameFields').props() as any).onChange(nameFields);
      });

      expect(onChange).toHaveBeenCalledWith(
        {
          name: {
            title,
            forename,
            middleName,
            surname
          },
          contactDetails: [
            {
              type: '' as ContactType,
              detail: ''
            }
          ],
          relationship
        },
        true
      );

      form.update();
      const type = ContactType.HomeNumber;
      const detail = '09483424555';
      const contactDetail = { type, detail };

      act(() => {
        (form.find('ContactDetailFields').props() as any).onChange(contactDetail);
      });

      expect(onChange).toHaveBeenCalledWith(
        {
          name: {
            title,
            forename,
            middleName,
            surname
          },
          contactDetails: [contactDetail],
          relationship
        },
        true
      );
    });
  });
});
