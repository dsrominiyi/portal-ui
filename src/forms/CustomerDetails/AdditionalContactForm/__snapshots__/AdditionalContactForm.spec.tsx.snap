// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalContactForm snapshots AdditionalContactForm matches snapshot 1`] = `
<div
  className="formSection "
  onKeyUp={[Function]}
  test-id="additional-contact-form"
>
  <QuestionBox
    inputFields={
      Array [
        <Dropdown
          attributes={
            Object {
              "test-id": "select-relationship",
            }
          }
          label="Relationship"
          onChange={[Function]}
          options={
            Array [
              "Spouse",
              "Sibling",
              "Parent",
              "Child",
              "Solicitor",
            ]
          }
          selected=""
        />,
      ]
    }
  />
  <SelectTitle
    onChange={[Function]}
    title=""
  />
  <NameFields
    fieldErrors={
      Object {
        "forename": undefined,
        "middleName": undefined,
        "surname": undefined,
      }
    }
    forename=""
    middleName=""
    onChange={[Function]}
    surname=""
  />
  <ContactDetailFields
    detail=""
    fieldErrors={
      Object {
        "detail": undefined,
        "type": undefined,
      }
    }
    onChange={[Function]}
    type=""
  />
</div>
`;

exports[`AdditionalContactForm snapshots AdditionalContactForm with initial values matches snapshot 1`] = `
<div
  className="formSection "
  onKeyUp={[Function]}
  test-id="additional-contact-form"
>
  <QuestionBox
    inputFields={
      Array [
        <Dropdown
          attributes={
            Object {
              "test-id": "select-relationship",
            }
          }
          label="Relationship"
          onChange={[Function]}
          options={
            Array [
              "Spouse",
              "Sibling",
              "Parent",
              "Child",
              "Solicitor",
            ]
          }
          selected=""
        />,
      ]
    }
  />
  <SelectTitle
    onChange={[Function]}
    title=""
  />
  <NameFields
    fieldErrors={
      Object {
        "forename": undefined,
        "middleName": undefined,
        "surname": undefined,
      }
    }
    forename=""
    middleName=""
    onChange={[Function]}
    surname=""
  />
  <ContactDetailFields
    detail=""
    fieldErrors={
      Object {
        "detail": undefined,
        "type": undefined,
      }
    }
    onChange={[Function]}
    type=""
  />
</div>
`;

exports[`AdditionalContactForm snapshots AdditionalContactForm without form errors matches snapshot 1`] = `
<div
  className="formSection "
  onKeyUp={[Function]}
  test-id="additional-contact-form"
>
  <QuestionBox
    inputFields={
      Array [
        <Dropdown
          attributes={
            Object {
              "test-id": "select-relationship",
            }
          }
          label="Relationship"
          onChange={[Function]}
          options={
            Array [
              "Spouse",
              "Sibling",
              "Parent",
              "Child",
              "Solicitor",
            ]
          }
          selected=""
        />,
      ]
    }
  />
  <SelectTitle
    onChange={[Function]}
    title=""
  />
  <NameFields
    fieldErrors={
      Object {
        "forename": undefined,
        "middleName": undefined,
        "surname": undefined,
      }
    }
    forename=""
    middleName=""
    onChange={[Function]}
    surname=""
  />
  <ContactDetailFields
    detail=""
    fieldErrors={
      Object {
        "detail": undefined,
        "type": undefined,
      }
    }
    onChange={[Function]}
    type=""
  />
</div>
`;
